File: CSharpCodeBase/aiactions/ai_movetoenemy.lua
local Class = require 'utils.hump.class'
local Timer = require 'utils.hump.timer'
local AINode = require 'utils.ailib.ainode'

-- Move to enemy
local AI_MoveToEnemy = Class({__includes=AINode, init = function (self, entity)
  AINode.init(self, entity)
  self.timer = Timer(self.entity.config:Get("chaseTimeout"))
end})

function AI_MoveToEnemy:Visit()
  if self.status == NODE_READY then
    self.status = NODE_RUNNING
    self.entity.mover:ResetSpeed()
    self.entity.mover:SetGoal(self.parent.selected_target, nil, self.entity.weaponContainer:GetAttackDistance())
    self.timer:Reset()
  elseif self.status == NODE_RUNNING then
    if not self.parent.selected_target.interactable then
      self.status = NODE_FAILURE
      return
    end
    if not self.entity.mover:IsHaveGoal() then
      self.status = NODE_SUCCESS
    else
      if self.timer:Tick() then
        self.status = NODE_FAILURE
      end
    end
  end
  return self.status
end

return AI_MoveToEnemy

File: CSharpCodeBase/aiactions/ai_movetoenemy.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class ai_movetoenemy :ainode{
 // Move to enemy
 public void init (self, entity);
   AINode.init(self, entity);
   this.timer = Timer(this.entity.config:Get("chaseTimeout"));
 }})
 public void Visit(){
   if(this.status == NODE_READY  ){ 
     this.status = NODE_RUNNING;
     this.entity.mover:ResetSpeed();
     this.entity.mover:SetGoal(this.parent.selected_target, null, this.entity.weaponContainer:GetAttackDistance());
     this.timer:Reset();
   }else{if this.status == NODE_RUNNING  ){ 
     if(not this.parent.selected_target.interactable  ){ 
       this.status = NODE_FAILURE;
       return;
     }
     if(not this.entity.mover:IsHaveGoal()  ){ 
       this.status = NODE_SUCCESS;
     }else{
       if(this.timer:Tick()  ){ 
         this.status = NODE_FAILURE;
       }
     }
   }
   return this.status;
 }
 }}